File: lang-core/src/test/java/com/jaitlapps/differ/impl/StringReaderTest.java
package com.jaitlapps.differ.impl;

import com.jaitlapps.differ.Reader;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class StringReaderTest {
    @Test
    public void testOneWord()
    {
        Reader wordReader = new StringReader("Третье");

        String word = wordReader.readNextWord();

        assertEquals("Третье", word);
        assertEquals(0, wordReader.getStartPosition());
        assertEquals(6, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testOneWordSpace()
    {
        Reader wordReader = new StringReader("   Третье      ");

        String word = wordReader.readNextWord();

        assertEquals("Третье", word);
        assertEquals(3, wordReader.getStartPosition());
        assertEquals(9, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testTwoWordPlus()
    {
        Reader wordReader = new StringReader("ere+err");

        String word = wordReader.readNextWord();

        assertEquals("ere", word);
        assertEquals(0, wordReader.getStartPosition());
        assertEquals(3, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("+", word);
        assertEquals(3, wordReader.getStartPosition());
        assertEquals(4, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("err", word);
        assertEquals(4, wordReader.getStartPosition());
        assertEquals(7, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    //@Test
    public void testTwoWordExpression()
    {
        Reader wordReader = new StringReader("qwe=123--4");

        String word = wordReader.readNextWord();

        assertEquals("qwe", word);
        assertEquals(0, wordReader.getStartPosition());
        assertEquals(3, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("=", word);
        assertEquals(3, wordReader.getStartPosition());
        assertEquals(4, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("123", word);
        assertEquals(4, wordReader.getStartPosition());
        assertEquals(7, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("-", word);
        assertEquals(7, wordReader.getStartPosition());
        assertEquals(8, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("-4", word);
        assertEquals(8, wordReader.getStartPosition());
        assertEquals(10, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testTwoWordExpression1()
    {
        Reader wordReader = new StringReader("qwe  =123-  4");

        String word = wordReader.readNextWord();

        assertEquals("qwe", word);
        assertEquals(0, wordReader.getStartPosition());
        assertEquals(3, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("=", word);
        assertEquals(5, wordReader.getStartPosition());
        assertEquals(6, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("123", word);
        assertEquals(6, wordReader.getStartPosition());
        assertEquals(9, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("-", word);
        assertEquals(9, wordReader.getStartPosition());
        assertEquals(10, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("4", word);
        assertEquals(12, wordReader.getStartPosition());
        assertEquals(13, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testThreeWord()
    {
        Reader wordReader = new StringReader("Третье 123 12345");

        String word = wordReader.readNextWord();

        assertEquals("Третье", word);
        assertEquals(0, wordReader.getStartPosition());
        assertEquals(6, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("123", word);
        assertEquals(7, wordReader.getStartPosition());
        assertEquals(10, wordReader.getEndPosition());

        word = wordReader.readNextWord();

        assertEquals("12345", word);
        assertEquals(11, wordReader.getStartPosition());
        assertEquals(16, wordReader.getEndPosition());

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }



    @Test
    public void testMinus1()
    {
        Reader wordReader = new StringReader("22+-4");

        String word = wordReader.readNextWord();

        assertEquals("22", word);

        word = wordReader.readNextWord();

        assertEquals("+", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("4", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testMinus2()
    {
        Reader wordReader = new StringReader("22- 4");

        String word = wordReader.readNextWord();

        assertEquals("22", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("4", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testMinus3()
    {
        Reader wordReader = new StringReader("22 - 4");

        String word = wordReader.readNextWord();

        assertEquals("22", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("4", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testMinus4()
    {
        Reader wordReader = new StringReader("22 -4");

        String word = wordReader.readNextWord();

        assertEquals("22", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("4", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testMinus5()
    {
        Reader wordReader = new StringReader("22-4");

        String word = wordReader.readNextWord();

        assertEquals("22", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("4", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }

    @Test
    public void testMinus6()
    {
        Reader wordReader = new StringReader("-");

        String word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }


    @Test
    public void testMinus7()
    {
        Reader wordReader = new StringReader("45 * 2 - 3");

        String word = wordReader.readNextWord();

        assertEquals("45", word);

        word = wordReader.readNextWord();

        assertEquals("*", word);

        word = wordReader.readNextWord();

        assertEquals("2", word);

        word = wordReader.readNextWord();

        assertEquals("-", word);

        word = wordReader.readNextWord();

        assertEquals("3", word);

        word = wordReader.readNextWord();
        assertEquals(null, word);
    }
}

File: lang-core/src/test/kotlin/com/jaitlapps/differ/reader/impl/StringReaderTest.kt
package com.jaitlapps.differ.reader.impl

import org.junit.Test
import kotlin.test.*

class StringReaderTest {
    @Test
    fun testOneWord() {
        val wordReader = StringReader("Третье")

        var word = wordReader.readNextWord()

        assertEquals("Третье", word?.lexeme)
        assertEquals(0, word?.startPosition)
        assertEquals(6, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testOneWordSpace() {
        val wordReader = StringReader("   Третье      ")

        var word = wordReader.readNextWord()

        assertEquals("Третье", word?.lexeme)
        assertEquals(3, word?.startPosition)
        assertEquals(9, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testTwoWordPlus() {
        val wordReader = StringReader("ere+err")

        var word = wordReader.readNextWord()

        assertEquals("ere", word?.lexeme)
        assertEquals(0, word?.startPosition)
        assertEquals(3, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("+", word?.lexeme)
        assertEquals(3, word?.startPosition)
        assertEquals(4, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("err", word?.lexeme)
        assertEquals(4, word?.startPosition)
        assertEquals(7, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    //@Test
    fun testTwoWordExpression() {
        val wordReader = StringReader("qwe=123--4")

        var word = wordReader.readNextWord()

        assertEquals("qwe", word?.lexeme)
        assertEquals(0, word?.startPosition)
        assertEquals(3, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("=", word?.lexeme)
        assertEquals(3, word?.startPosition)
        assertEquals(4, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("123", word?.lexeme)
        assertEquals(4, word?.startPosition)
        assertEquals(7, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)
        assertEquals(7, word?.startPosition)
        assertEquals(8, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("-4", word?.lexeme)
        assertEquals(8, word?.startPosition)
        assertEquals(10, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testTwoWordExpression1() {
        val wordReader = StringReader("qwe  =123-  4")

        var word = wordReader.readNextWord()

        assertEquals("qwe", word?.lexeme)
        assertEquals(0, word?.startPosition)
        assertEquals(3, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("=", word?.lexeme)
        assertEquals(5, word?.startPosition)
        assertEquals(6, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("123", word?.lexeme)
        assertEquals(6, word?.startPosition)
        assertEquals(9, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)
        assertEquals(9, word?.startPosition)
        assertEquals(10, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)
        assertEquals(12, word?.startPosition)
        assertEquals(13, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testThreeWord() {
        val wordReader = StringReader("Третье 123 12345")

        var word = wordReader.readNextWord()

        assertEquals("Третье", word?.lexeme)
        assertEquals(0, word?.startPosition)
        assertEquals(6, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("123", word?.lexeme)
        assertEquals(7, word?.startPosition)
        assertEquals(10, word?.endPosition)

        word = wordReader.readNextWord()

        assertEquals("12345", word?.lexeme)
        assertEquals(11, word?.startPosition)
        assertEquals(16, word?.endPosition)

        word = wordReader.readNextWord()
        assertNull(word)
    }


    @Test
    fun testMinus1() {
        val wordReader = StringReader("22+-4")

        var word = wordReader.readNextWord()

        assertEquals("22", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("+", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testMinus2() {
        val wordReader = StringReader("22- 4")

        var word = wordReader.readNextWord()

        assertEquals("22", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testMinus3() {
        val wordReader = StringReader("22 - 4")

        var word = wordReader.readNextWord()

        assertEquals("22", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testMinus4() {
        val wordReader = StringReader("22 -4")

        var word = wordReader.readNextWord()

        assertEquals("22", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testMinus5() {
        val wordReader = StringReader("22-4")

        var word = wordReader.readNextWord()

        assertEquals("22", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("4", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }

    @Test
    fun testMinus6() {
        val wordReader = StringReader("-")

        var word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }


    @Test
    fun testMinus7() {
        val wordReader = StringReader("45 * 2 - 3")

        var word = wordReader.readNextWord()

        assertEquals("45", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("*", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("2", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("-", word?.lexeme)

        word = wordReader.readNextWord()

        assertEquals("3", word?.lexeme)

        word = wordReader.readNextWord()
        assertNull(word)
    }
}